Fix ORTTrainer failure on DeBERTa(base/v2/sew_d) fp16 training

JingyaHuang opened the pull request:
# What does this PR do?

__Context__

It was reported in optimum https://github.com/huggingface/optimum/issues/305 that the training on DeBERTa with optimum.onnxruntime.ORTTrainer is broken. 
After investigation, the break comes from two causes:
* At that time `XDropOut` didn't have a symbolic function. And it has been implemented by @garymm in https://github.com/huggingface/transformers/pull/17502 and has been merged to the main of transformers.
* The implementation of DeBERTa have some numpy/math operations that led to incorrect export. This will be fixed in https://github.com/huggingface/transformers/pull/18272. 

However with those two fixes, the fp32 training will work, but the mixed-precision training will fail due to mismatched inputs dtype for some `Matmul` nodes. In https://github.com/huggingface/transformers/pull/18272, some `sqrt` results are cast to `fp32`, and they need to be re-casted to fp16 before `Matmul` ops, and this PR is supposed to add the re-cast part.

Fixes #https://github.com/huggingface/optimum/issues/305

## Who can review?

@LysandreJik @patrickvonplaten @lewtun

JingyaHuang committed:
Modify DeBERTa-v2 modeling

JingyaHuang committed:
Modify DeBERTa-v2 modeling

JingyaHuang committed:
Modify DeBERTa-v2 modeling

JingyaHuang committed:
Partly fix DeBERTa

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
revert v2

JingyaHuang committed:
revert v2

JingyaHuang committed:
revert v2

JingyaHuang committed:
revert v2

JingyaHuang committed:
revert v2

JingyaHuang committed:
Fix DeBERTa-v2

JingyaHuang committed:
Fix DeBERTa-v2

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Fix Deberta with PR in trfrs

JingyaHuang committed:
Fix Deberta with PR in trfrs

JingyaHuang committed:
Fix Deberta with PR in trfrs

JingyaHuang committed:
Fix Deberta with PR in trfrs

JingyaHuang committed:
Fix Deberta with PR in trfrs

JingyaHuang committed:
add xdropout symbolic

JingyaHuang committed:
add xdropout symbolic

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Fix docstring

JingyaHuang committed:
Fix sew_d

JingyaHuang committed:
Correct dtype inside torch.sqrt

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Try pos_key_layer dtype

JingyaHuang committed:
Try pos_query_layer dtype

JingyaHuang committed:
Try query_layer dtype

JingyaHuang committed:
Try query_layer dtype

JingyaHuang committed:
Try query_layer dtype

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
Revert matmul changes

JingyaHuang committed:
Hacky test torch.half

JingyaHuang committed:
Test remove dtype

JingyaHuang committed:
Test matmul unified dtype

JingyaHuang committed:
Test matmul unified dtype

JingyaHuang committed:
Test p2c  matmul unified dtype

JingyaHuang committed:
Test p2c matmul unified dtype

JingyaHuang committed:
Revert test c2p matmul unified dtype

JingyaHuang committed:
Revert test c2p matmul unified dtype

JingyaHuang committed:
c2p matmul fixed

JingyaHuang committed:
Test dtype change attention_scores

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

JingyaHuang committed:
DeBERTa fixed!

JingyaHuang committed:
Test optimized fix

JingyaHuang committed:
Test optimized fix

JingyaHuang committed:
Confirm optimized fix for DeBERTa

JingyaHuang committed:
Fix sew_d

JingyaHuang committed:
Fix sew_d

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18529). All of your documentation changes will be reflected on that endpoint.

JingyaHuang committed:
Fix copies of sew-d

JingyaHuang committed:
Fix copies of sew-d

JingyaHuang committed:
Merge branch 'huggingface:main' into fix-deberta-tracing

Narsil committed:
Adding a new `align_to_words` param to qa pipeline. (#18010)

* Adding a new `align_to_words` param to qa pipeline.

* Update src/transformers/pipelines/question_answering.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Import protection.

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stevhliu committed:
📝 update metric with evaluate (#18535)

YouJiacheng committed:
Restore _init_weights value in no_init_weights (#18504)

* Recover _init_weights value in no_init_weights

For potential nested use. 
In addition, users might modify private no_init_weights as well.

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Remove private variable change check

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger committed:
Clean up comment

stevhliu committed:
📝 update documentation build section (#18548)

younesbelkada committed:
`bitsandbytes` - `Linear8bitLt` integration into `transformers` models (#17901)

* first commit

* correct replace function

* add final changes

- works like charm!
- cannot implement tests yet
- tested

* clean up a bit

* add bitsandbytes dependencies

* working version

- added import function
- added bitsandbytes utils file

* small fix

* small fix

- fix import issue

* fix import issues

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* refactor a bit

- move bitsandbytes utils to utils
- change comments on functions

* reformat docstring

- reformat docstring on init_empty_weights_8bit

* Update src/transformers/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* revert bad formatting

* change to bitsandbytes

* refactor a bit

- remove init8bit since it is useless

* more refactoring

- fixed init empty weights issue
- added threshold param

* small hack to make it work

* Update src/transformers/modeling_utils.py

* Update src/transformers/modeling_utils.py

* revmoe the small hack

* modify utils file

* make style + refactor a bit

* create correctly device map

* add correct dtype for device map creation

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* apply suggestions

- remove with torch.grad
- do not rely on Python bool magic!

* add docstring

 - add docstring for new kwargs

* add docstring

- comment `replace_8bit_linear` function
- fix weird formatting

* - added more documentation
- added new utility function for memory footprint tracking
- colab demo to add

* few modifs

- typo doc
- force cast into float16 when load_in_8bit is enabled

* added colab link

* add test architecture + docstring a bit

* refactor a bit testing class

* make style + refactor a bit

* enhance checks

- add more checks
- start writing saving test

* clean up a bit

* male style

* add more details on doc

* add more tests

- still needs to fix 2 tests

* replace by "or"

- could not fix it from GitHub GUI

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* refactor a bit testing code + add readme

* make style

* fix import issue

* Update src/transformers/modeling_utils.py

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

* add few comments

* add more doctring + make style

* more docstring

* raise error when loaded in 8bit

* make style

* add warning if loaded on CPU

* add small sanity check

* fix small comment

* add bitsandbytes on dockerfile

* Improve documentation

- improve documentation from comments

* add few comments

* slow tests pass on the VM but not on the CI VM

* Fix merge conflict

* make style

* another test should pass on a multi gpu setup

* fix bad import in testing file

* Fix slow tests

- remove dummy batches
- no more CUDA illegal memory errors

* odify dockerfile

* Update docs/source/en/main_classes/model.mdx

* Update Dockerfile

* Update model.mdx

* Update Dockerfile

* Apply suggestions from code review

* few modifications

- lm head can stay on disk/cpu
- change model name so that test pass

* change test value

- change test value to the correct output
- torch bmm changed to baddmm in bloom modeling when merging

* modify installation guidelines

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* replace `n`by `name`

* merge `load_in_8bit` and `low_cpu_mem_usage`

* first try - keep the lm head in full precision

* better check

- check the attribute `base_model_prefix` instead of computing the number of parameters

* added more tests

* Update src/transformers/utils/bitsandbytes.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

* improve documentation

- fix typos for installation
- change title in the documentation

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

gante committed:
TF: XLA-trainable DeBERTa v2 (#18546)

* fix deberta issues

* add different code paths for gpu and tpu

* shorter gpu take along axis

* Stable Dropout without tf cond

* variable must be float

sgugger committed:
Preserve hub-related kwargs in AutoModel.from_pretrained (#18545)

* Preserve hub-related kwargs in AutoModel.from_pretrained

* Fix tests

* Remove debug statement

Rocketknight1 committed:
TF Examples Rewrite (#18451)

* Finished QA example

* Dodge a merge conflict

* Update text classification and LM examples

* Update NER example

* New Keras metrics WIP, fix NER example

* Update NER example

* Update MC, summarization and translation examples

* Add XLA warnings when shapes are variable

* Make sure batch_size is consistently scaled by num_replicas

* Add PushToHubCallback to all models

* Add docs links for KerasMetricCallback

* Add docs links for prepare_tf_dataset and jit_compile

* Correct inferred model names

* Don't assume the dataset has 'lang'

* Don't assume the dataset has 'lang'

* Write metrics in text classification

* Add 'framework' to TrainingArguments and TFTrainingArguments

* Export metrics in all examples and add tests

* Fix training args for Flax

* Update command line args for translation test

* make fixup

* Fix accidentally running other tests in fp16

* Remove do_train/do_eval from run_clm.py

* Remove do_train/do_eval from run_mlm.py

* Add tensorflow tests to circleci

* Fix circleci

* Update examples/tensorflow/language-modeling/run_mlm.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Update examples/tensorflow/test_tensorflow_examples.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Update examples/tensorflow/translation/run_translation.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Update examples/tensorflow/token-classification/run_ner.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Fix save path for tests

* Fix some model card kwargs

* Explain the magical -1000

* Actually enable tests this time

* Skip text classification PR until we fix shape inference

* make fixup

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

sgugger committed:
Use commit hash to look in cache instead of calling head (#18534)

* Use commit hash to look in cache instead of calling head

* Add tests

* Add attr for local configs too

* Stupid typos

* Fix tests

* Update src/transformers/utils/hub.py

Co-authored-by: Julien Chaumond <julien@huggingface.co>

* Address Julien's comments

Co-authored-by: Julien Chaumond <julien@huggingface.co>

julien-c committed:
`pipeline` support for `device="mps"` (or any other string) (#18494)

* `pipeline` support for `device="mps"` (or any other string)

* Simplify `if` nesting

* Update src/transformers/pipelines/base.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Fix? @sgugger

* passing `attr=None` is not the same as not passing `attr` 🤯

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stevhliu committed:
Update philosophy to include other preprocessing classes (#18550)

* 📝 update philosophy to include other preprocessing classes

* 🖍 apply feedbacks

sgugger committed:
Properly move cache when it is not in default path (#18563)

unography committed:
Adds CLIP to models exportable with ONNX (#18515)

* onnx config for clip

* default opset as 14

* changes from the original repo

* input values order fix

* outputs fix

* remove unused import

* ran make fix-copies

* black format

* review comments: forward ref, import fix, model change revert, .to cleanup

* make style

* formatting fixes

* revert groupvit

* comment for cast to int32

* comment fix

* make .T as .t() for onnx conversion

* ran make fix-copies

* remove unneeded comment

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* fix copies

* remove comment

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ydshieh committed:
raise atol for MT5OnnxConfig (#18560)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

mrwyattii committed:
fix string (#18568)

joihn committed:
Segformer TF: fix output size in documentation (#18572)

* Segformer TF: fix output size in doc

* Segformer pytorch: fix output size in doc

Co-authored-by: Maxime Gardoni <maxime.gardoni@ecorobotix.com>

alaradirik committed:
Fix resizing bug in OWL-ViT (#18573)

* Fixes resizing bug in OWL-ViT
* Defaults to square resize if size is set to an int
* Sets do_center_crop default value to False

pocca2048 committed:
Fix LayoutLMv3 documentation (#17932)

* fix typos

* fix sequence_length docs of LayoutLMv3Model

* delete trailing white spaces

* fix layoutlmv3 docs more

* apply make fixup & quality

* change to two versions of input docstring

* apply make fixup & quality

sgugger committed:
Skip broken tests

donebydan committed:
Change BartLearnedPositionalEmbedding's forward method signature to support Opacus training (#18486)

* changing BartLearnedPositionalEmbedding forward signature and references to it

* removing debugging dead code (thanks style checker)

* blackened modeling_bart file

* removing copy inconsistencies via make fix-copies

* changing references to copied signatures in Bart variants

* make fix-copies once more

* using expand over repeat (thanks @michaelbenayoun)

* expand instead of repeat for all model copies

Co-authored-by: Daniel Jones <jonesdaniel@microsoft.com>

flozi00 committed:
german docs translation (#18544)

* Create _config.py

* Create _toctree.yml

* Create index.mdx

not sure about "du / ihr" oder "sie"

* Create quicktour.mdx

* Update _toctree.yml

* Update build_documentation.yml

* Update build_pr_documentation.yml

* fix build

* Update index.mdx

* Update quicktour.mdx

* Create installation.mdx

* Update _toctree.yml

iiLaurens committed:
Deberta V2: Fix critical trace warnings to allow ONNX export (#18272)

* Fix critical trace warnings to allow ONNX export

* Force input to `sqrt` to be float type

* Cleanup code

* Remove unused import statement

* Update model sew

* Small refactor

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

* Use broadcasting instead of repeat

* Implement suggestion

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

* Match deberta v2 changes in sew_d

* Improve code quality

* Update code quality

* Consistency of small refactor

* Match changes in sew_d

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

JingyaHuang committed:
Update DeBERTa fix

JingyaHuang committed:
Update DeBERTa fix

JingyaHuang commented:
close as it turned to be too messy even after rebasing.
